pysteps/nowcasts/steps.py

```diff
@@ -33,6 +33,873 @@
+                n = mask_f * self.timestep / self.kmperpixel
+                self.struct = iterate_structure(self.struct, int((n - 1) / 2.0))
+                # Compute and apply the dilated mask for each ensemble member
+                self.mask_precip = nowcast_utils.compute_dilated_mask(
+                    self.mask_precip, self.struct, self.mask_rim
+                )
+                self.mask_precip = [
+                    self.mask_precip.copy() for _ in range(self.n_ens_members)
+                ]
+        else:
+            self.mask_precip = None
+
+        if self.noise_method is None and self.precip_mask is None:
+            self.precip_mask = [
+                self.precip_cascades[0][i].copy() for i in range(self.n_cascade_levels)
+            ]
+        print("Precipitation mask initialized successfully.")
+
+    def _initialize_fft_objects(self):
+        """
+        Initialize FFT objects for each ensemble member.
+        """
+        self.fft_objs = []
+        for _ in range(self.n_ens_members):
+            fft_obj = utils.get_method(self.fft_method, shape=self.precip.shape[1:])
+            self.fft_objs.append(fft_obj)
+        print("FFT objects initialized successfully.")
+
+    def _initialize_state(self):
+        """
+        Initialize the state dictionary used during the nowcast iteration.
+        """
+        return {
+            "fft_objs": self.fft_objs,
+            "mask_prec": self.mask_precip,
+            "precip_cascades": self.precip_cascades,
+            "precip_decomp": self.precip_decomposed,
+            "precip_m": self.precip_mask,
+            "precip_m_d": self.precip_mask_decomposed,
+            "randgen_prec": self.random_generator_precip,
+        }
+
+    def _initialize_params(self, precip):
+        """
+        Initialize the params dictionary used during the nowcast iteration.
+        """
+        return {
+            "decomp_method": self.decomposition_method,
+            "domain": self.domain,
+            "domain_mask": self.domain_mask,
+            "filter": self.bandpass_filter,
+            "fft": self.fft,
+            "generate_noise": self.generate_noise,
+            "mask_method": self.mask_method,
+            "mask_rim": self.mask_rim,
+            "mu_0": self.mu_0,
+            "n_cascade_levels": self.n_cascade_levels,
+            "n_ens_members": self.n_ens_members,
+            "noise_method": self.noise_method,
+            "noise_std_coeffs": self.noise_std_coeffs,
+            "num_ensemble_workers": self.num_ensemble_workers,
+            "phi": self.phi,
+            "pert_gen": self.perturbation_generator,
+            "probmatching_method": self.probmatching_method,
+            "precip": precip,
+            "precip_thr": self.precip_threshold,
+            "recomp_method": self.recomp_method,
+            "struct": self.struct,
+            "war": self.war,
+        }
+
+    def _update_state(self, state, params):
```

Comment:
The `state` could also be a dataclass. That way you could explicitly define all variables that are in the state.